code/game/objects/machinery/miner.dm

```diff
@@ -43,21 +50,90 @@
+
+/obj/machinery/miner/proc/attempt_upgrade(obj/item/minerupgrade/upgrade, mob/user, params)
+	if(miner_upgrade_type)
+		return FALSE
+	if(user.skills.getRating("construction")<SKILL_CONSTRUCTION_ADVANCED)
+		to_chat(user, "<span class='info'>You can't figure out how to install the complex module.</span>")
+		return FALSE
+	user.visible_message("<span class='notice'>[user] begins attaching a module to [src]'s sockets.</span>")
+	to_chat(user, "<span class='info'>You begin installing the [upgrade] on the miner.</span>")
+	if(!do_after(user, 15 SECONDS, TRUE, src, BUSY_ICON_BUILD))
+		return FALSE
+	if(!upgrade.use())
+		return FALSE
+	var/choice = upgrade.uptype
+	switch(choice)
+		if(MINER_RESISTANT)
+			max_miner_integrity = 300
+			miner_integrity = 300
+		if(MINER_COMPACTOR)
+			if(mineral_produced == /obj/structure/ore_box/platinum)
+				mineral_produced = /obj/item/compactorebox/platinum
+			else
+				mineral_produced = /obj/item/compactorebox/phoron
+		if(MINER_OVERCLOCKED)
+			required_ticks = 35
 
+	miner_upgrade_type = choice
+	qdel(upgrade)
+	user.visible_message("<span class='notice'>[user] attaches the [choice] to the [src]!</span>")
+	playsound(loc,'sound/items/screwdriver.ogg', 25, TRUE)
+	update_icon()
+
+/obj/machinery/miner/attackby(obj/item/I,mob/user,params)
+	. = ..()
+
+	if(istype(I,/obj/item/minerupgrade))
+		var/obj/item/minerupgrade/upgrade = I
+		if(!(miner_status == MINER_RUNNING))
+			to_chat(user, "<span class='info'>[src]'s module sockets seem bolted down.</span>")
+			return FALSE
+		src.attempt_upgrade(upgrade,user)
 
 /obj/machinery/miner/welder_act(mob/living/user, obj/item/I)
 	. = ..()
+	var/obj/item/tool/weldingtool/weldingtool = I
+	if((miner_status == MINER_RUNNING)&&miner_upgrade_type)
+		if(!weldingtool.remove_fuel(2,user))
+			to_chat(user, "<span class='info'>You need more welding fuel to complete this task!</span>")
+			return FALSE
+		to_chat(user, "<span class='info'>You begin uninstalling the [miner_upgrade_type] from the miner</span>")
+		user.visible_message("<span class='notice'>[user] begins dismantling the [miner_upgrade_type] from the miner</span>")
+		if(!do_after(user, 30 SECONDS, TRUE, src, BUSY_ICON_BUILD))
+			return FALSE
+		user.visible_message("<span class='notice'>[user] dismantles the [miner_upgrade_type] from the miner</span>")
+		var/obj/item/minerupgrade/upgrade
+		switch(miner_upgrade_type)
+			if(MINER_RESISTANT)
+				upgrade = new /obj/item/minerupgrade/reinforcement
+			if(MINER_OVERCLOCKED)
+				upgrade = new /obj/item/minerupgrade/overclock
+			if(MINER_COMPACTOR)
+				upgrade = new /obj/item/minerupgrade/compactor
+		upgrade.loc = user.loc
+		miner_upgrade_type = ""
+		max_miner_integrity = 100
+		miner_integrity = 100
+		if(mineral_produced == /obj/item/compactorebox/phoron)
+			mineral_produced = /obj/structure/ore_box/phoron
+		else
+			mineral_produced = /obj/structure/ore_box/platinum
+		required_ticks = 70
+		update_icon()
```

Comment:
```suggestion
		switch(miner_upgrade_type)
			if(MINER_RESISTANT)
				new /obj/item/minerupgrade/reinforcement
				max_miner_integrity = 100
				miner_integrity = 100
			if(MINER_OVERCLOCKED)
				new /obj/item/minerupgrade/overclock
				required_ticks = 70
			if(MINER_COMPACTOR)
				new /obj/item/minerupgrade/compactor
				if(istype(mineral_produced, /obj/item/compactorebox/phoron))
					mineral_produced = /obj/structure/ore_box/phoron
				else
					mineral_produced = /obj/structure/ore_box/platinum
		miner_upgrade_type = null
		update_icon()
```
Couple things to note here:
A) we use forceMove instead of setting loc directly
B) Setting to null is better than setting to an empty string
C) Stuff should be handled inside the switch for the state of posterity
D) I didnt do it in this suggestion, but you should be using 
`var/myvar = initial(myvar)`
here instead of setting numbers directly